### ISSUE TITLE
 MongoError: E11000 duplicate key error index:

### ISSUE THREAD

**Issue by pawankorotane** [balderdashy/sails#2591]
Hi Guys,

Could you please help me out solving this issue I am getting the user location info data before it creates the record.

It creates the record but with no location data in it then I am getting duplicate key error I thing Mongodb create a data before it get the location data.

I don't know how to fix this issue I am new to Node.js thanks.

/**
 * Users.js
 *
 * @description :: TODO: You might write a short summary of how this model works and what it represents here.
 * @docs        :: http://sailsjs.org/#!documentation/models
 */

var bcrypt = require('bcrypt');
var crypto =  require('crypto');
var request =  require('request'); 
module.exports = {
    autosubscribe: ['destroy', 'update'],
    attributes: {
        username: {
            type: 'string',
            required: true,
            regex : /^\w+$/,
            unique: true,
            maxLength: 30,
            minLength : 5,
        },
        email : {
            type : 'email',
            unique : true,
            required : true,
            maxLength: 30,
        },
        age : {
            type : 'integer',
            required : true,
            min : 18,
            integer : true
        },
        password: {
            type: 'string',
            required: true,
            minLength : 5,
            regex: /^(?=.*\d)(?=.*[a-z])(?=.*[A-Z]).{6,}$/

        },
        active : {
            type : 'boolean',
            defaultsTo : false
           
        },
        key : {
            type: 'string',
            defaultsTo : crypto.randomBytes(20).toString('hex')
        },
        key_time : {
            type : 'datetime',
            defaultsTo : new Date().toISOString()
            
        },

        online : "boolean",
        token : {
            collection : 'tokens',
            via : 'user_id'
            
        },
    rooms: {
        type : 'array',
        required : true
    },
    cordinates: {
       type: 'json' // index:'2dsphere', unique : false
    },
    country : {
        type : 'string'
    },
    city : {
        type : 'string'
    },
    ip:{
        type : 'string',
       // ip : true
    },
    time_zone : {
        type : 'string'
    },

    toJSON: function() {        // remove the password field from rest api
            var obj = this.toObject();
            delete obj.password;
           // delete obj.key;
            delete obj.token;
            return obj;
        }
    },
    validationMessages: {
        username: {
          required: 'Username is required',
          maxLength : 'Username can not be greater than 30 characters',
          minLength : 'Username should be at least 5 characters long',
          unique : 'User with this username name is already exist in the system please select another username',
          regex : 'Username must contain only letters, numbers and underscores!'
        },
        email: {
          required : 'Email is required',
          email : 'Please enter valid email',
          maxLength : 'Email can not be greater than 50 characters'
            },
        age :{
            required : 'Age is required',
            min : "Minimum age of allowed is 18",
            integer : "Age needs to be in numeric"
        },
        password : {
            required : 'Password field is required',
            minLength : 'Password should be at least 5 characters long',
            regex : 'Password must contain at least 6 characters, including UPPER/lowercase and numbers'
        },
        rooms : {
            required : "Please select at least one room"
        }
    },
    beforeCreate: function(values, cb) {
         
         check_password(values,cb);
         Users.findOneByEmail(values.email).exec(function(err, user) {
              if (user){
                return cb({"error":"Email is already exist in the system."});
               
              }
                
        });

         // check user name exist in the system or not
         Users.findOneByUsername(values.username).exec(function(err, user) {
              if (user){
                return cb({"error":"Username is already exist."});
                
              }
                
        });

        get_location(values, cb);

        
    },
    afterCreate: function(values,  cb, req){
        console.log(values);
        delete values.userEmail

         var activate_url  = sails.getBaseurl() + '/#/account/'+values.key+'/'+values.id;
              var message =  '<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">'+
                                '<html><head><title>Welcome to'+ sails.config.email.application_name +'!</title></head>'+
                                '<body><div style="max-width: 800px; margin: 0; padding: 30px 0;"><table width="80%" border="0" cellpadding="0" cellspacing="0"><tr>'+
                                '<td width="5%"></td>'+
                                '<td align="left" width="95%" style="font: 13px/18px Arial, Helvetica, sans-serif;">'+
                        '<h2 style="font: normal 20px/23px Arial, Helvetica, sans-serif; margin: 0; padding: 0 0 18px; color: black;">Welcome to'+sails.config.email.application_name+'!</h2>'+
                        'Thanks for joining '+sails.config.email.application_name+'. We listed your sign in details below, make sure you keep them safe.<br />'+
                        'To verify your email address, please follow this link:<br /><br /><big style="font: 16px/18px Arial, Helvetica, sans-serif;"><b><a href="'+activate_url+'" style="color: #3366cc;">Finish your registration...</a></b></big><br />'+
                        '<br />Link does not work? Copy the following link to your browser address bar:<br /><nobr><a href="'+activate_url+'" style="color: #3366cc;">'+activate_url+'</a></nobr><br />'+
                        '<br />Please verify your email within 36 hours, otherwise your registration will become invalid and you will have to do the same process again.<br /><br /><br />'+
                        'Your username: '+values.username+'<br />'+
                        'Your email address: '+values.email+'<br />'+
                        '<br /><br />'+
                        'Have fun!<br />'+
                        'The '+sails.config.email.application_name+' Team'+
                        '</td></tr></table></div></body></html>';

        EmailService.sendEmail({username : values.username, email : values.email, subject : 'Account Activation Email', message : message});

        cb();
        
        
    }
    
};

    /**
        Checks the passwrod and compare it as well
    **/

    function check_password(values,cb){

        // Encrypt password
        if(values.password != ''){
            bcrypt.hash(values.password, 10, function(err, hash) {
            if (err)
                return cb(err);
            values.password = hash;
            //calling cb() with an argument returns an error. Useful for canceling the entire operation if some criteria fails.
            cb();
        });

        delete values.passwordConfirm;   // delete the password confirm
        }
        

        
    }

function get_location(values, cb){
    request('http://freegeoip.net/json/'+values.ip, function (error, response, body) {
          console.log(body);
          if (!error && response.statusCode == 200) {
            
            values.cordinates  = {latitude: body.latitude,longitude : body.longitude};
            if(typeof body.country !== undefined){
                values.country  =  body.country_name;
            }

            if(typeof body.city !== undefined){
                values.city  =  body.city;
            }

            if(typeof body.ip !== undefined){
                values.ip  =  body.ip;
            }

            if(typeof body.time_zone !== undefined){
                values.time_zone  =  body.time_zone;
            }
           
          }else{
              return cb({error : "location not found"})
          }
        });
    return cb();
}

**Comment by CWyrtzen** [comment]
@pawankorotane with all subsequent releases are you still having this issue?

**Comment by sailsbot** [comment]
Thanks for posting, @pawankorotane.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!
